lib/transaction/priority_queue.go

```diff
@@ -139,6 +142,35 @@ func (spq *PriorityQueue) Push(txn *ValidTransaction) (common.Hash, error) {
 	return hash, nil
 }
 
+// PopWithTimer returns the next valid transaction from the queue.
+// When the timer expires, it returns `nil`.
+func (spq *PriorityQueue) PopWithTimer(timer *time.Timer) (transaction *ValidTransaction) {
```

Comment:
:warning: Check if the queue has a transaction first to avoid spinning an extra goroutine AND waiting spq.pollInterval even a transaction is already present in the queue.

```suggestion
func (spq *PriorityQueue) PopWithTimer(timer *time.Timer) (transaction *ValidTransaction) {
	transaction = spq.Pop()
	if transaction != nil {
		return transaction
	}

```